### PR TITLE
[mkFit] Mid-way support for Phase2 initialStep in CMSSW

### PR DESCRIPTION
**Pull request by osschar**
### PR description:

This PR adds the ability to run mkFit on Phase-2 data / geometry. Only initialStep iteration is supported so far, with a dedicated workflow (`24834.702`).

* Add CMSSW Phase-2 hit and seed converters.

* Properly support binary readout for Phase-2 strips where cluster-charge cut can
not be applied.

* Improve standalone performance monitoring & debugging tools.

**It requires also a new JSON file in cmssw/data repository:**
https://github.com/trackreco/RecoTracker-MkFit/pull/3

### PR validation:
In TTbar events, with PU:
- Phase-2 (PU200, where mkFit is only used for initialStep track building): http://uaf-10.t2.ucsd.edu/~mmasciov/MIC/Phase2/MTV_TTbarPU200_mcRun4_withTiming/
- Phase-1 (2024; no change, as expected): http://uaf-10.t2.ucsd.edu/~mmasciov/MIC/Phase2/MTV_TTbarPU_Run3_2024/

NOTE: closed in favor of PR #146.

### CODE REVIEW

File: RecoTracker/MkFit/plugins/MkFitSeedConverter.cc [hunk 1]
```diff
@@ -147,7 +147,7 @@ mkfit::TrackVec MkFitSeedConverter::convertSeeds(const edm::View<TrajectorySeed>
         LogTrace("MkFitSeedConverter") << " adding hit detid " << detId.rawId() << " index " << clusterRef.index()
                                        << " ilay " << ilay;
         ret.back().addHitIdx(clusterRef.index(), ilay, 0);  // per-hit chi2 is not known
-      } else {
+      } else if (!mkFitGeom.isPhase2()) {                   // XXXX why is !phase2 if needed here?
```
Comment (slava77) [inline RecoTracker/MkFit/plugins/MkFitSeedConverter.cc:150]: to be resolved before submission. I think there should at least be an "else" with a throw for the remaining option

File: RecoTracker/MkFitCore/interface/IdxChi2List.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    int nholes;           // number of holes (used for sorting)
+    float pt;             // pt (used for sorting)
+    float chi2;           // total chi2 (used for sorting)
+    float chi2_hit;       // chi2 of the added hit
+    float score;          // score used for candidate ranking
+
+    IdxChi2List() = default;
+    IdxChi2List& operator=(const IdxChi2List&) = default;
+
+    IdxChi2List(unsigned int) { reset(); }
```
Comment (slava77) [inline RecoTracker/MkFitCore/interface/IdxChi2List.h:23]: the `int` argument is confusing.
Isn't it more clear to just add default initialization above ` float score = 0;` ?

Comment (osschar) [reply RecoTracker/MkFitCore/interface/IdxChi2List.h:24]: During tracking we create the struct on the stack and then manually assign to every member.

There was a second usage when dumping hit-match-info to a TTree for NN cand-score training. There, in creation of the exported structure in selectHitIndicies() we do not know yet if the given hit will be selected -- so we pre-initialize the struct to all zeroes there (as we might not get to process it later).

I have committed a comment above the unsigned int constructor to clarify this.

File: RecoTracker/MkFit/plugins/MkFitSiStripHitConverter.cc [hunk 3]
```diff
@@ -61,6 +73,8 @@ class MkFitSiStripHitConverter : public edm::global::EDProducer<> {
 MkFitSiStripHitConverter::MkFitSiStripHitConverter(edm::ParameterSet const& iConfig)
     : stripRphiRecHitToken_{consumes<SiStripRecHit2DCollection>(iConfig.getParameter<edm::InputTag>("rphiHits"))},
       stripStereoRecHitToken_{consumes<SiStripRecHit2DCollection>(iConfig.getParameter<edm::InputTag>("stereoHits"))},
+      siPhase2RecHitToken_{
+          consumes<Phase2TrackerRecHit1DCollectionNew>(iConfig.getParameter<edm::InputTag>("siPhase2Hits"))},
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitSiStripHitConverter.cc]: Done.

File: RecoTracker/MkFit/plugins/MkFitSiStripHitConverter.cc [hunk 4]
```diff
@@ -76,6 +90,7 @@ void MkFitSiStripHitConverter::fillDescriptions(edm::ConfigurationDescriptions&
 
   desc.add("rphiHits", edm::InputTag{"siStripMatchedRecHits", "rphiRecHit"});
   desc.add("stereoHits", edm::InputTag{"siStripMatchedRecHits", "stereoRecHit"});
+  desc.add("siPhase2Hits", edm::InputTag{"siPhase2RecHits"});
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitSiStripHitConverter.cc]: Moved to dedicated plugin.

File: RecoTracker/IterativeTracking/python/InitialStep_cff.py [hunk 5]
```diff
@@ -444,10 +445,14 @@
...
 InitialStep = cms.Sequence(InitialStepTask)
 
+from RecoLocalTracker.Phase2TrackerRecHits.Phase2TrackerRecHits_cfi import siPhase2RecHits
 from Configuration.ProcessModifiers.trackingMkFitCommon_cff import trackingMkFitCommon
 _InitialStepTask_trackingMkFitCommon = InitialStepTask.copy()
+_InitialStepTask_trackingMkFitCommon_Phase2 = InitialStepTask.copy()
 _InitialStepTask_trackingMkFitCommon.add(mkFitSiPixelHits, mkFitEventOfHits, mkFitGeometryESProducer)
-trackingMkFitCommon.toReplaceWith(InitialStepTask, _InitialStepTask_trackingMkFitCommon)
+_InitialStepTask_trackingMkFitCommon_Phase2.add(siPhase2RecHits, mkFitSiPixelHits, mkFitSiPhase2Hits, mkFitEventOfHits, mkFitGeometryESProducer)
+(trackingMkFitCommon & (~trackingPhase2PU140)).toReplaceWith(InitialStepTask, _InitialStepTask_trackingMkFitCommon_Phase2)
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/IterativeTracking/python/InitialStep_cff.py]: Fixed.

File: RecoTracker/MkFit/plugins/MkFitSeedConverter.cc (hunk 1 above)
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitSeedConverter.cc]: Removed.

File: RecoTracker/IterativeTracking/python/InitialStep_cff.py [hunk 6]
```diff
@@ -240,6 +240,7 @@
 import RecoTracker.MkFit.mkFitProducer_cfi as mkFitProducer_cfi
 import RecoTracker.MkFit.mkFitOutputConverter_cfi as mkFitOutputConverter_cfi
 mkFitSiPixelHits = mkFitSiPixelHitConverter_cfi.mkFitSiPixelHitConverter.clone() # TODO: figure out better place for this module?
+mkFitSiPhase2Hits = mkFitSiStripHitConverter_cfi.mkFitSiStripHitConverter.clone(rphiHits=cms.InputTag(""), stereoHits=cms.InputTag(""))
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/IterativeTracking/python/InitialStep_cff.py]: Done.

File: RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc [hunk 7]
```diff
@@ -72,7 +72,7 @@ MkFitEventOfHitsProducer::MkFitEventOfHitsProducer(edm::ParameterSet const& iCon
   if (usePixelQualityDB_) {
     pixelQualityToken_ = esConsumes();
   }
-  if (useStripStripQualityDB_) {
+  if (useStripStripQualityDB_ && !((iConfig.getParameter<edm::InputTag>("stripHits"))==edm::InputTag{"mkFitSiPhase2Hits"})) {
     stripQualityToken_ = esConsumes();
```
[the remark this answers is not in the document]
Comment (mmasciov) [reply RecoTracker/MkFit/plugins/MkFitEventOfHitsProducer.cc]: Done.

File: Configuration/PyReleaseValidation/python/relval_2026.py [hunk 8]
```diff
@@ -27,6 +27,7 @@
 numWFIB.extend([24834.0,24834.911,24834.103]) #2026D98 DDD XML, DD4hep XML, aging
 numWFIB.extend([25061.97]) #2026D98 premixing stage1 (NuGun+PU)
 numWFIB.extend([24834.5,24834.9]) #2026D98 pixelTrackingOnly, vector hits
+numWFIB.extend([24834.72026]) #2026D98 mkFit tracking (initialStep)
```
Comment (slava77) [inline Configuration/PyReleaseValidation/python/relval_2026.py:30]: can this be just .7?
"2026" in 72026 is a bit of a dead end (can't be similarly usefully extended to add slight similar/modified; unclear if it makes it easier to remember).
The neighboring tracking wf is `.701`; just the next one will be `.702`

**Comment by slava77** [comment]
there is a merge conflict here now. ` Configuration/PyReleaseValidation/python/relval_2026.py `
is it just a matter of resyncing the fork?

File: Configuration/PyReleaseValidation/python/relval_2026.py (hunk 8 above)
Comment (slava77) [reply Configuration/PyReleaseValidation/python/relval_2026.py:30]: > can this be just .7?

I was looking at the implementation and it seems like having conditionals inside of the old workflow is not possible. 
OTOH, it's not obvious that two workflows with the same offset value is impossible

**Comment by mmasciov** [comment]
> there is a merge conflict here now. `Configuration/PyReleaseValidation/python/relval_2026.py` is it just a matter of resyncing the fork?

Solved.

File: Configuration/PyReleaseValidation/python/relval_2026.py (hunk 8 above)
Comment (mmasciov) [reply Configuration/PyReleaseValidation/python/relval_2026.py]: `.7` is identified as duplicate. Changed to `.702`

**Comment by mmasciov** [comment]
Closing in favor of 'compressed' PR #146.

**Pull request closed without merging**